src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs

```diff
@@ -517,6 +517,19 @@ public T WithTelemetry(ITelemetryConfig telemetryConfig)
             return (T)this;
         }
 
+        /// <summary>
+        /// Configures MSAL to retry on 5xx server errors. When enabled (on by default), MSAL will wait 1 second after recieving a 5xx error and then retry the
+        /// http request again.
+        /// </summary>
+        /// <param name="shouldRetry">Determines if MSAL should retry on 5xx errors.</param>
+        /// <returns></returns>
+        public T WithRetryOnServerErrors(bool shouldRetry)
```

Comment:
@jmprieur for new public API review

Reply:
I'm feelin like it would be good to spec out the whole retry feature so that we do this holistically. We discussed having some retry adapter or something that maybe user can pass.
Or I was thinking something like `.WithRetryOptions(retryOptions)`, then `RetryOptions` can have something like `RetryOptions.Retry5xx`, `RetryOptions.Retry418`, `RetryOptions.RetryTaskCancellationEx`, `RetryOptions.RetryDelay`, `RetryOptions.RetryMaxCount`.

I understand it'll take more time, but I think it'll help us deliver a  better feature holistically.

Reply:
Yes, I was thinking about a more general-purpose API as well, which could include cache retries, retries on HTTP errors etc. At the same time, it's better if retries were handled externally, where specialized libraries like Poly can provide the golden standard of retries, i.e. exponential backoff. 

Alternative to not add a new API which we might never extend, is to reuse `WithHttpClientFactory(IHttpClientFactory factory, RetryOptions retryOptions)`.

Thoughts? 

CC: @jmprieur and @rymeskar

Reply:
Here's an idea based on feedback from @mjcheetham 

```csharp

WithRetryStrategy(IHttpRetryStrategy httpRetryStrategy);

// Return null to not retry. Return a timestap to retry after waiting for it.
interface IHttpRetryStrategy
{
    // Invoked when the HTTP response is 4xx or 5xx. 
    TimeStamp? OnHttpResponseFailed(HttpResponse httpResponse4xx5xx);

    // Invoked on exceptions
    TimeStamp? OnHttpException(Exception ex);

```

Reply:
I think we want to avoid providing retries on all 4xx since ESTS already has a mechanism for 425 with the retry after header that the user is expected to handle. Seems like the new implementations may potentially class with it.

Reply:
It's a fair point, we need to document 4xx and especially 429 TooManyRetries

Alternative proposal from @mjcheetham  https://gist.github.com/mjcheetham/44880df497c147b1f90a73fe86f8a3b7

Reply:
Discussed this with @jmprieur as well, and the perspective is:

While we could add a complex extensiblity point such as the one described [here](https://gist.github.com/mjcheetham/44880df497c147b1f90a73fe86f8a3b7), we recognize that retry libraries like Poly work perfectly fine at HttpClient level and Msal already exposes this layer. So the proposal is to add an overload to `WithHttpClientFactory`:

```WithHttpClientFactory(IMsalHttpClientFactory factory, bool disableDefaultRetryPolicy)```

Reply:
It sounds right to simplify our API and let retry libraries do the job. But we need to document this well as we are getting several questions regarding it.

Reply:
Can you also provide an MSAL extensibility point (separate package) where you would provide the default poly implementation/configuration of what you think are the right values?

Reply:
You are probably in a better position to advise about good retry policies @rymeskar, since you work with high scale services. We'd take your advice as recommendation for others.